src/coreclr/jit/morph.cpp

```diff
@@ -13886,6 +13886,144 @@ GenTree* Compiler::fgOptimizeBitwiseAnd(GenTreeOp* andOp)
     return nullptr;
 }
 
+//------------------------------------------------------------------------
+// fgOptimizeRelationalComparisonWithCasts: optimizes a comparison operation.
+//   Recognizes comparisons against various cast operands and tries to remove
+//   them. E.g.:
+//
+//   *  GE        int
+//   +--*  CAST      long <- ulong <- uint
+//   |  \--*  X         int
+//   \--*  CNS_INT   long
+//
+//   to:
+//
+//   *  GE_un     int
+//   +--*  X         int
+//   \--*  CNS_INT   int
+//
+//   same for:
+//
+//   *  GE        int
+//   +--*  CAST      long <- ulong <- uint
+//   |  \--*  X         int
+//   \--*  CAST      long <- [u]long <- int
+//      \--*  ARR_LEN   int
+//
+// Arguments:
+//   tree - the GT_LE/GT_LT/GT_GE/GT_GT tree to morph.
+//
+// Return Value:
+//   Returns the same tree where operands might have narrower types
+//
+GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* tree)
+{
+    assert(tree->OperIs(GT_LE, GT_LT, GT_GE, GT_GT));
+
+    GenTree* castedOp        = tree->gtGetOp1();
+    GenTree* knownPositiveOp = tree->gtGetOp2();
+
+    // Caller is expected to call this function only if we have CAST nodes
+    assert(castedOp->OperIs(GT_CAST) || knownPositiveOp->OperIs(GT_CAST));
+
+    if (optValnumCSE_phase || gtIsActiveCSE_Candidate(tree) || gtIsActiveCSE_Candidate(castedOp) ||
+        gtIsActiveCSE_Candidate(knownPositiveOp))
+    {
+        // We're going to modify all of them
+        return tree;
+    }
+
+    if (!castedOp->TypeIs(TYP_LONG))
+    {
+        // We can extend this logic to handle small types as well, but currently it's done mostly to
+        // assist range check elimination
+        return tree;
+    }
+
+    bool knownPositiveIsOp2;
+    if (knownPositiveOp->IsIntegralConst() ||
+        ((knownPositiveOp->OperIs(GT_CAST) && knownPositiveOp->gtGetOp1()->OperIs(GT_ARR_LENGTH))))
```

Comment:
Is it possible to use `GenTree::IsNeverNegative` here?

Reply:
Yeah, this PR was filed before it appeared 🙂

Reply:
Turns out I need to keep it as - I need exactly either const or (cast)Arr_Len hee, more over IsNeverNegative return false for (sign_extent)ArrLen (can be fixed as a follow up)

Reply:
Just to understand it, why can we not do the optimization for other positive operands?

Reply:
Honestly, this method is mostly to catch a specific popular pattern related to bound checks, to handle more cases (and more types, not just INT) it has to be rewritten to something more complicated/advanced .e.g "demanded bits" in LLVM. I'll look separately if it worth doing by estimating potential benefits